sycl/doc/extensions/proposed/sycl_ext_oneapi_prefetch.asciidoc

```diff
@@ -0,0 +1,437 @@
+feedback.  Interfaces defined in this specification may not be implemented yet
+or may be in a preliminary state.  The specification itself may also change in
+incompatible ways before it is finalized.  *Shipping software products should
+not rely on APIs defined in this specification.*
+
+
+== Overview
+
+Many devices targeted by SYCL support software prefetch operations, which
+act as asynchronous memory reads intended to populate cache(s). Prefetches are
+generally used to hide memory latency, and their deployment may be a critical
+component for software tuning on some hardware.
+
+The `multi_ptr::prefetch` function provided by SYCL 2020 is insufficient to
+cover many real use-cases -- it does not acknowledge the existence of a cache
+hierarchy, and does not provide a mechanism for cooperative (i.e. group)
+prefetches.
+
+This proposal addresses these shortcomings by introducing a set of
+free functions that groups of work-items can use to prefetch data into specific
+levels of cache (controlled via compile-time properties).
+
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.  An implementation supporting this extension must predefine the
+macro `SYCL_EXT_ONEAPI_PREFETCH` to one of the values defined in the table
+below.  Applications can test for the existence of this macro to determine if
+the implementation supports this feature, or applications can test the macro's
+value to determine which of the extension's features the implementation
+supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|The APIs of this experimental extension are not versioned, so the
+ feature-test macro always has this value.
+|===
+
+
+=== Specifying cache levels
+
+The level of cache targeted by a prefetch is specified using compile-time
+properties. This extension defines hints for four levels of cache,
+corresponding to the four levels of cache currently defined in SYCL (e.g.
+as used by `info::partition_affinity_domain`).
+
+[NOTE]
+====
+Not all devices targeted by SYCL will have four levels of cache. Some of these
+hints may have no meaning on some devices. However, implementations are
+encouraged to map these constants to the closest level of cache available.
+====
+
+When multiple cache levels are specified, the lowest level takes precedence
+(e.g. a request to prefetch into L1 and L4 is treated as a request to prefetch
+into L1). When no cache levels are specified, the default behavior is to
+prefetch into the lowest level cache (i.e. L1).
+
+[NOTE]
+====
+Future hints may alter the default behavior of prefetches with respect to cache
+levels. Any such alterations are expected to be documented in the definition
+of those new hints.
+====
+
+[source,c++]
+----
+namespace sycl::ext::oneapi::experimental {
+
+enum class cache_level {
+  L1,
+  L2,
+  L3,
+  L4,
+};
+
+struct nontemporal;
```

Comment:
by only having `nontemporal` defined here, are we implicitly saying that without this being specified it is by default `temporal` cache messages?

Reply:
More or less.  I think without `nontemporal` it's really just saying something like "prefetch into this level of cache, with the default hardware behavior". That's _probably_ temporal, but I don't think it's required to be. Adding `nontemporal` is an additional hint that the data won't be reused.

I realized I didn't add any documentation for what `nontemporal` means, so I've added a brief description in c88269af0080.